zipline/assets/asset_writer.py

```
@@ -0,0 +1,779 @@
+            A connection to our SQLite database.
+        fuzzy_char: string
+            A string to be used in fuzzy matching.
+        allow_sid_assignment: boolean
+            If True, allow the writer to assign sids where necessary.
+        constraints: boolean
+            If True, add SQL constraints to tables.
+        """
+
+        self.conn = db_conn
+
+        self.fuzzy_char = fuzzy_char
+        self.allow_sid_assignment = allow_sid_assignment
+
+        # This flag controls if the AssetDBWriter is allowed to generate its
+        # own sids. If False, metadata that does not contain a sid will raise
+        # an exception when building assets.
+        if allow_sid_assignment:
+            self.end_date_to_assign = normalize_date(
+                pd.Timestamp('now', tz='UTC'))
+
+        # Create SQL tables.
+        self.init_db(self.conn, constraints)
+
+        # Write to SQL tables.
+        for sid, metadata in self.load_data(self._data):
+            self.write_block(sid, **metadata)
+
+    def write_block(self, identifier, **kwargs):
+        """
+        Inserts the given metadata kwargs to the entry for the given
+        sid. Matching fields in the existing entry will be overwritten.
+        Will be deprecated in future versions of zipline.
+        """
+
+        if identifier in self.metadata_cache:
+            # Multiple pass insertion no longer supported.
+            # This could and probably should raise an Exception, but is
+            # currently just a short-circuit for compatibility with existing
+            # testing structure in the test_algorithm module which creates
+            # multiple sources which all insert redundant metadata.
+            return
+
+        entry = {}
+
+        for key, value in kwargs.items():
+            # Do not accept invalid fields
+            if key not in ASSET_FIELDS:
+                continue
+            # Do not accept Nones
+            if value is None:
```

Should these fail silently?